### PR TITLE
Feat. TE7-T123 Action Controller Overview

### PR DESCRIPTION
**Pull request by Dhyey54**
Action Controller Overview Practical:

Create the user controller and add custom events for login and logout Set session on successful login and end session on logout Create a car controller with all possible actions(index, new, etc) Create a new action called search and use it to find a matching car name Use strong params inside controller actions to see what error arises if the strong parameter is not used, add SS of error. The user should be able to access the cars list if logged in Redirect users to the sign-in page if the user is not logged in. Use filters on some actions with only/except
Use flash messages to notify users about success/failure response The user should be able to download the pdf file. Put some content into that pdf file. Practice setting and getting cookies
Use various authentications for some actions for the controller: http_basic_authenticate_with Use  rescue_from to show the message "Record is not found" if an exception  ActiveRecord::RecordNotFound occurs

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -33,6 +33,9 @@ gem "jbuilder"
 
 gem "byebug"
 
+gem "prawn"
```
Comment (poonamsimform) [inline Gemfile:37]: Please add a comment for why we are using this gem.

File: app/controllers/cars_controller.rb [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  before_action :authenticate
+  before_action :require_login
+  before_action :admin_login, only: [:new, :edit, :destroy]
+
+  def index
+    @cars = Car.all.order(id: :asc)
+  end
+
+  def show
+    @car = Car.find(params[:id])
```
Comment (poonamsimform) [inline app/controllers/cars_controller.rb]: Let's now use `filters` for this find method. To DRY the code.

File: app/views/users/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,50 @@
+<div class="mx-auto text-center">
+
+  <% if @current_user != nil %>
+    <% # Code for Successful Logged in user%>
```
Comment (poonamsimform) [inline app/views/users/index.html.erb]: Please check this document [https://medium.com/@mjagiello/commenting-in-embedded-ruby-files-11a9966d2c0d](url) for how to use comments.

File: app/views/shared/_navbar.html.erb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<nav class="navbar navbar-expand-lg bg-secondary navbar-dark">
+  <div class="container-fluid justify-content-start">
+    <a class="navbar-brand" href="/home">Home</a>
+    <a class="navbar-brand" href="/products">Products</a>
+    <a class="navbar-brand" href="/authors">Authors</a>
+    <a class="navbar-brand" href="/books">Books</a>
+    <a class="navbar-brand" href="/students">Students</a>
+    <a class="navbar-brand" href="/faculties">Faculties</a>
+    <a class="navbar-brand" href="/users">Users</a>
+    <a class="navbar-brand" href="/cars">Cars</a>
```
Comment (poonamsimform) [inline app/views/shared/_navbar.html.erb]: Please use `link_to`

File: app/views/users/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        <th>Username</th>
+      </tr>
+      <% @users.each do |user| %>
+        <tr>
+          <td><%= user.id %></td>
+          <td><%= user.username %></td>
+        </tr>
+      <% end %>
+    </table>
+  <%= link_to 'Cars', cars_path, class: "text-decoration-none text-dark btn btn-info" %>
```
Comment (poonamsimform) [inline app/views/users/index.html.erb]: Please remove this. This is not needed now as we have made a header

File: app/views/cars/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,37 @@
+<%= render "shared/logout" %>
```
Comment (poonamsimform) [inline app/views/cars/_form.html.erb]: No need to write this partial in every file. Just put it into `application.html.erb` file

File: app/views/cars/_search.html.erb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<%= render "shared/logout" %>
```
Comment (poonamsimform) [inline app/views/cars/_search.html.erb]: Please remove

File: app/views/cars/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,27 @@
+<h1 class="text-center my-4">Cars</h1>
+<%= render "search" %>
+<%= render "shared/logout" %>
```
Comment (poonamsimform) [inline app/views/cars/index.html.erb]: Please remove

File: app/views/shared/_logout.html.erb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<% @current_user = User.where(id: session[:user_id]).first %>
```
Comment (poonamsimform) [inline app/views/shared/_logout.html.erb]: Do we need to find the `current_user` here? OR Have we already access the `@current_user`

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
+class UsersController < ApplicationController
+  http_basic_authenticate_with name: "admin", password: "123"
+  
+  def index
+    current_user
+    @users = User.all.order(id: :asc)
+  end
+
+  def create
+    @user = User.where(username: params[:username], password: params[:password]).first
```
Comment (poonamsimform) [inline app/controllers/users_controller.rb]: No, need to define instance variable( `@` ) if you are using that variable in same method.

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    if @user
+      session[:user_id] = @user[:id]
+      cookies[:user_name] = @user[:username]
```
Comment (poonamsimform) [inline app/controllers/users_controller.rb]: What is both of these variables do?

File: app/views/cars/search.html.erb [hunk 12]
```diff
@@ -0,0 +1,22 @@
+<h1 class="text-center my-4">Cars</h1>
```
Comment (poonamsimform) [inline app/views/cars/search.html.erb:1]: @Dhyey54 When are we using this functionality?

Comment (Dhyey54) [reply app/views/cars/search.html.erb:1]: We search then the search method of the cars controller is called and the page is viewed

File: app/controllers/cars_controller.rb [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+  before_action :authenticate
+  before_action :admin_login, only: [:new, :edit, :destroy]
+  before_action :find_car, only: [:show, :edit, :update, :destroy]
+
+  def index
+    @cars = Car.all.order(id: :asc)
+  end
+
+  def show
```
Comment (poonamsimform) [inline app/controllers/cars_controller.rb]: Please go through this document to check how to define empty body methods - [https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/EmptyMethod](url)

File: app/controllers/cars_controller.rb [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  def new
+    @car = Car.new
+  end
+
+  def edit
```
Comment (poonamsimform) [inline app/controllers/cars_controller.rb]: same here.

**Review by poonamsimform — APPROVED** [review]

**Pull request merged**